### DISCUSSION TITLE
Current Typestates support?

### DISCUSSION THREAD

**Discussion by dan-f** [statelyai/xstate#3706] [Q&A] [answered]
Hi there,

I'm working with xstate + typescript. I'm not totally sure from your [docs](https://xstate.js.org/docs/guides/typescript.html) as to whether the recommended approach is to pass generic type parameters to `createMachine` (including [typestates](https://xstate.js.org/docs/guides/typescript.html)) or to make use of typegen. Right now I'm using typegen.

I specifically would like to be able to associate Context subtypes with particular states, which I believe was the point of typestates. Is this supported with typegen? This bit from the docs seems to indicate so...:

> ...all state.matches types are type-safe. Enums are currently not supported by our static analysis tool. It's also unlikely that we'll ever support them with typegen due to the complexity they add for comparatively little gain.
>
> Instead of enums, use typegen and rely on the strength of the type-safety it provides.

...but it's unclear what is specifically meant by "all state.matches types are type-safe". Typescript is aware of the legal state names as an input to `state.matches('myState')`. But there is no type narrowing over `state.context`, which is the point of typestates.

Here's how I'm accessing these APIs, in case it's something on my end:
- starting the service via `@xstate/react` with `useInterpret(machine, options)`
- then calling e.g. `service.subscribe((state) => { if (state.matches('myState') { /* no narrowing */ } })`

Thanks!

**Comment by Andarist** [comment, marked as answer]
The recommended approach is to use typegen. 

> I specifically would like to be able to associate Context subtypes with particular states, which I believe was the point of typestates.

Yes, that's what typestates were about but they were never validated so they were never really typesafe. We'd like to explore typesafe ways of doing this kind of thing in the future but it is not a trivial task and we don't actually have concrete ideas on how to support this with a proper type-level support. It's on our minds though.

**Reply by dan-f** [reply]
Thanks for the reply. Just to confirm, what I'm hearing here is that aside from typestates (which is not recommended), there is no current support for associating context subtypes with particular states. Is that correct? And where is that roughly at on the roadmap?

**Reply by Andarist** [reply]
> there is no current support for associating context subtypes with particular states. Is that correct? 

Yes, this is correct.

> And where is that roughly at on the roadmap?

I'd like to revisit this after releasing v5 - so I'd say that the typestate feature won't happen any time "soon". If you'd have any ideas on how this could be implemented - I would happily brainstorm some possibilities.

**Reply by dan-f** [reply]
Great, super clear. I don't have time to contribute for the foreseeable future sadly, but will let you know if that changes!
